fpga_interchange/lookahead.cc

```diff
@@ -0,0 +1,1522 @@
+ *  copyright notice and this permission notice appear in all copies.
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ *  WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ *  MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ *  ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ *  WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ *  ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ *  OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *
+ */
+
+#include "lookahead.h"
+
+#include <boost/filesystem.hpp>
+#include <capnp/message.h>
+#include <capnp/serialize.h>
+#include <kj/filesystem.h>
+#include <kj/std/iostream.h>
+#include <queue>
+#include <sstream>
+#include <zlib.h>
+
+#include "context.h"
+#include "flat_wire_map.h"
+#include "log.h"
+#include "sampler.h"
+#include "scope_lock.h"
+
+#if defined(NEXTPNR_USE_TBB)
+#include <tbb/parallel_for_each.h>
+#endif
+
+NEXTPNR_NAMESPACE_BEGIN
+
+static constexpr size_t kNumberSamples = 4;
+static constexpr int32_t kMaxExploreDist = 20;
+
+// Initial only explore with a depth of this.
+static constexpr int32_t kInitialExploreDepth = 30;
+
+struct RoutingNode
+{
+    WireId wire_to_expand;
+    delay_t cost;
+    int32_t depth;
+
+    bool operator<(const RoutingNode &other) const { return cost < other.cost; }
+};
+
+struct PipAndCost
+{
+    PipId upstream_pip;
+    delay_t cost_from_src;
+    int32_t depth;
+};
+
+static void expand_input(const Context *ctx, WireId input_wire, HashTables::HashMap<TypeWireId, delay_t> *input_costs)
+{
+    HashTables::HashSet<WireId> seen;
+    std::priority_queue<RoutingNode> to_expand;
+
+    RoutingNode initial;
+    initial.cost = 0;
+    initial.wire_to_expand = input_wire;
+
+    to_expand.push(initial);
+
+    while (!to_expand.empty()) {
+        RoutingNode node = to_expand.top();
+        to_expand.pop();
+
+        auto result = seen.emplace(node.wire_to_expand);
+        if (!result.second) {
+            // We've already done an expansion at this wire.
+            continue;
+        }
+
+        for (PipId pip : ctx->getPipsUphill(node.wire_to_expand)) {
+            if (ctx->is_pip_synthetic(pip)) {
+                continue;
+            }
+            WireId new_wire = ctx->getPipSrcWire(pip);
+            if (new_wire == WireId()) {
+                continue;
+            }
+
+            RoutingNode next_node;
+            next_node.wire_to_expand = new_wire;
+            next_node.cost = node.cost + ctx->getPipDelay(pip).maxDelay() + ctx->getWireDelay(new_wire).maxDelay();
+
+            if (ctx->is_site_port(pip)) {
+                // Done with expansion, record the path if cheaper.
+                // Only the first path to each wire will be the cheapest.
+                TypeWireId route_to;
+                route_to.type = ctx->chip_info->tiles[pip.tile].type;
```

Comment:
is there a reason to be reaching deep into `chip_info` here rather than using the `TypeWireId` constructor like elsewhere?

Reply:
Yes, because if `src_index` is a node, we don't want `TypeWireId` to be the node, we want it to be the local tile wire connected to the source of the pip.

Reply:
Same story with line 201.

Reply:
Fixed